File: codex-next/scripts/auto-recover-stop.py
#!/usr/bin/env python3
from __future__ import annotations

import hashlib
import json
import os
import sys
from pathlib import Path
from typing import Any
import re


if sys.platform.startswith("win"):
    import io as _io

    for _stream_name in ("stdin", "stdout", "stderr"):
        _stream = getattr(sys, _stream_name, None)
        if _stream is None:
            continue
        if hasattr(_stream, "reconfigure"):
            try:
                _stream.reconfigure(encoding="utf-8", errors="replace")  # type: ignore[union-attr]
            except Exception:
                pass
        elif hasattr(_stream, "detach"):
            try:
                setattr(sys, _stream_name, _io.TextIOWrapper(_stream.detach(), encoding="utf-8", errors="replace"))
            except Exception:
                pass


MAX_ATTEMPTS_TOTAL = 5
MAX_ATTEMPTS_BY_KIND = {
    "transient_rate_limit": 3,
    "transient_overload": 2,
    "usage_limit": 2,
}

CONTINUE_MESSAGES = {
    "transient_rate_limit": (
        "The previous turn appears to have been interrupted by an OpenAI rate limit. "
        "Continue exactly where you left off and do not restart from scratch."
    ),
    "transient_overload": (
        "The previous turn appears to have been interrupted by a temporary OpenAI server overload. "
        "Continue exactly where you left off and do not restart from scratch."
    ),
    "usage_limit": (
        "The previous turn appears to have been interrupted by a Codex model or usage limit. "
        "Continue exactly where you left off if the limit has cleared, and do not restart from scratch."
    ),
}

STOP_MESSAGES = {
    "transient_rate_limit": (
        "Auto-retry stopped after repeated rate-limit interruptions. Check `/status`, wait for the limit window to "
        "reset, then continue manually."
    ),
    "transient_overload": (
        "Auto-retry stopped after repeated temporary OpenAI overload responses. Check `/status` and try again later, "
        "then continue manually."
    ),
    "usage_limit": (
        "Auto-retry stopped after repeated Codex model or usage-limit interruptions. Check `/status`, wait for reset, "
        "add credits if applicable, or switch to a lower-cost model, then continue manually."
    ),
}

ERROR_RE = re.compile(
    r"\b(error|failed|failure|interrupted|aborted|stopped|unable|cannot|can't|timed out|timeout)\b",
    re.IGNORECASE,
)
RATE_LIMIT_RE = re.compile(
    r"(\b429\b|rate[\s_-]?limit|too many requests|rate[\s_-]?limited|retry[\s_-]?after)",
    re.IGNORECASE,
)
OVERLOAD_RE = re.compile(
    r"(\b503\b|service unavailable|server error|internal server error|bad gateway|gateway timeout|overloaded|temporarily unavailable)",
    re.IGNORECASE,
)
USAGE_LIMIT_RE = re.compile(
    r"(usage limit|quota exceeded|credits?(?: exhausted| depleted)?|workspace usage limit|model (?:cap|limit)|monthly budget|spending limit)",
    re.IGNORECASE,
)
USAGE_CONTEXT_RE = re.compile(
    r"\b(reached|hit|exceeded|exhausted|depleted|reset|wait|later|remaining|quota|limit)\b",
    re.IGNORECASE,
)


def plugin_root(script_path: Path | None = None) -> Path:
    base = script_path or Path(__file__).resolve()
    return base.parents[1]


def default_state_dir(root: Path | None = None) -> Path:
    plugin_data = os.environ.get("PLUGIN_DATA")
    if plugin_data:
        return Path(plugin_data)
    return (root or plugin_root()) / ".local-state"


def read_stdin() -> str:
    try:
        return sys.stdin.read()
    except Exception:
        return ""


def parse_payload(raw: str) -> dict[str, Any]:
    if not raw.strip():
        return {}
    try:
        payload = json.loads(raw)
    except json.JSONDecodeError:
        return {}
    return payload if isinstance(payload, dict) else {}


def stable_session_key(payload: dict[str, Any]) -> str:
    transcript_path = payload.get("transcript_path")
    if isinstance(transcript_path, str) and transcript_path:
        return "transcript-" + hashlib.sha256(transcript_path.encode("utf-8")).hexdigest()[:16]

    cwd = payload.get("cwd") or ""
    model = payload.get("model") or ""
    key = f"{cwd}::{model}"
    return "fallback-" + hashlib.sha256(key.encode("utf-8")).hexdigest()[:16]


def state_file_path(payload: dict[str, Any], state_dir: Path | None = None) -> Path:
    base = state_dir or default_state_dir()
    return base / f"{stable_session_key(payload)}.json"


def load_state(path: Path) -> dict[str, Any]:
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except Exception:
        data = {}

    if not isinstance(data, dict):
        data = {}

    attempts = data.get("attempts_by_kind")
    if not isinstance(attempts, dict):
        attempts = {}

    return {
        "offset": int(data.get("offset", 0) or 0),
        "attempts_total": int(data.get("attempts_total", 0) or 0),
        "attempts_by_kind": {
            "transient_rate_limit": int(attempts.get("transient_rate_limit", 0) or 0),
            "transient_overload": int(attempts.get("transient_overload", 0) or 0),
            "usage_limit": int(attempts.get("usage_limit", 0) or 0),
        },
        "last_processed_turn_id": str(data.get("last_processed_turn_id", "") or ""),
        "transcript_prefix_hash": str(data.get("transcript_prefix_hash", "") or ""),
    }


def save_state(path: Path, state: dict[str, Any]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_text(json.dumps(state, ensure_ascii=False, indent=2), encoding="utf-8")


def read_transcript_delta(
    transcript_path: str | None,
    offset: int,
    known_prefix_hash: str = "",
) -> tuple[str, int, str]:
    if not transcript_path:
        return "", offset, ""

    path = Path(transcript_path)
    if not path.exists():
        return "", offset, ""

    try:
        size = path.stat().st_size
    except OSError:
        return "", offset, ""

    try:
        with path.open("rb") as fh:
            prefix_bytes = fh.read(128)
    except OSError:
        return "", offset, ""

    prefix_hash = hashlib.sha256(prefix_bytes).hexdigest() if prefix_bytes else ""
    rotated = bool(known_prefix_hash and prefix_hash and prefix_hash != known_prefix_hash)
    start = 0 if rotated or size < offset else offset
    if size == start:
        return "", size, prefix_hash

    try:
        with path.open("rb") as fh:
            fh.seek(start)
            chunk = fh.read(size - start)
    except OSError:
        return "", offset, prefix_hash

    text = chunk.decode("utf-8", errors="replace")
    return text, size, prefix_hash


def build_signal_text(payload: dict[str, Any], state: dict[str, Any]) -> tuple[str, int, str]:
    parts: list[str] = []
    message = payload.get("last_assistant_message")
    if isinstance(message, str) and message:
        parts.append(message)

    transcript_path = payload.get("transcript_path")
    chunk, next_offset, prefix_hash = read_transcript_delta(
        transcript_path if isinstance(transcript_path, str) else None,
        int(state.get("offset", 0) or 0),
        str(state.get("transcript_prefix_hash", "") or ""),
    )
    if chunk:
        parts.append(chunk)

    return "\n".join(parts), next_offset, prefix_hash


def classify_text(text: str) -> str | None:
    if not text:
        return None

    if USAGE_LIMIT_RE.search(text) and (ERROR_RE.search(text) or USAGE_CONTEXT_RE.search(text)):
        return "usage_limit"
    if RATE_LIMIT_RE.search(text) and ERROR_RE.search(text):
        return "transient_rate_limit"
    if OVERLOAD_RE.search(text) and ERROR_RE.search(text):
        return "transient_overload"
    return None


def continue_response(kind: str) -> dict[str, Any]:
    return {"decision": "block", "reason": CONTINUE_MESSAGES[kind]}


def stop_response(kind: str, state: dict[str, Any]) -> dict[str, Any]:
    used = int(state["attempts_by_kind"].get(kind, 0))
    cap = MAX_ATTEMPTS_BY_KIND[kind]
    detail = f"{kind} retries exhausted ({used}/{cap}; total {state['attempts_total']}/{MAX_ATTEMPTS_TOTAL})."
    return {
        "continue": False,
        "stopReason": detail,
        "systemMessage": STOP_MESSAGES[kind],
    }


def should_skip_turn(payload: dict[str, Any], state: dict[str, Any]) -> bool:
    if payload.get("stop_hook_active") is True:
        return True

    turn_id = payload.get("turn_id")
    if isinstance(turn_id, str) and turn_id and turn_id == state.get("last_processed_turn_id", ""):
        return True

    return False


def process_stop(payload: dict[str, Any], state_dir: Path | None = None) -> dict[str, Any] | None:
    state_path = state_file_path(payload, state_dir=state_dir)
    state = load_state(state_path)

    if should_skip_turn(payload, state):
        return None

    text, next_offset, prefix_hash = build_signal_text(payload, state)
    state["offset"] = next_offset
    state["transcript_prefix_hash"] = prefix_hash

    turn_id = payload.get("turn_id")
    if isinstance(turn_id, str) and turn_id:
        state["last_processed_turn_id"] = turn_id

    kind = classify_text(text)
    if kind is None:
        save_state(state_path, state)
        return None

    kind_attempts = int(state["attempts_by_kind"].get(kind, 0))
    total_attempts = int(state.get("attempts_total", 0))
    if kind_attempts >= MAX_ATTEMPTS_BY_KIND[kind] or total_attempts >= MAX_ATTEMPTS_TOTAL:
        save_state(state_path, state)
        return stop_response(kind, state)

    state["attempts_by_kind"][kind] = kind_attempts + 1
    state["attempts_total"] = total_attempts + 1
    save_state(state_path, state)
    return continue_response(kind)


def emit_json(payload: dict[str, Any]) -> None:
    sys.stdout.write(json.dumps(payload, ensure_ascii=False))


def main() -> int:
    payload = parse_payload(read_stdin())
    if not payload:
        return 0

    response = process_stop(payload)
    if response is not None:
        emit_json(response)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: codex-next/scripts/auto-recover-stop.mjs
import crypto from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";
import {
  appendStopEvent,
  ensureHookState,
  openAnalyticsDatabase,
  queryHookState,
  recordHookState,
  stableSessionKey,
} from "./lib/analytics-store.mjs";
import {
  CONTINUE_MESSAGES,
  MAX_ATTEMPTS_BY_KIND,
  MAX_ATTEMPTS_TOTAL,
  STOP_MESSAGES,
  classifyStopText,
} from "./lib/classify-stop.mjs";

const MODULE_DIR = path.dirname(fileURLToPath(import.meta.url));
const WINDOWS_UTF8_DONE = process.platform === "win32";

function safeReconfigure(streamName) {
  const stream = process[streamName];
  if (!stream) {
    return;
  }

  if (typeof stream.reconfigure === "function") {
    try {
      stream.reconfigure({ encoding: "utf8", errors: "replace" });
    } catch {
      // ignore
    }
  }
}

if (WINDOWS_UTF8_DONE) {
  safeReconfigure("stdin");
  safeReconfigure("stdout");
  safeReconfigure("stderr");
}

function readStdin() {
  try {
    return fs.readFileSync(0, "utf8");
  } catch {
    return "";
  }
}

function parsePayload(raw) {
  if (typeof raw !== "string" || raw.trim().length === 0) {
    return null;
  }

  try {
    const parsed = JSON.parse(raw);
    return parsed && typeof parsed === "object" && !Array.isArray(parsed) ? parsed : null;
  } catch {
    return null;
  }
}

function isString(value) {
  return typeof value === "string" && value.length > 0;
}

function defaultString(value) {
  return isString(value) ? value : "";
}

function readTranscriptDelta(transcriptPath, offset, knownPrefixHash = "") {
  if (!isString(transcriptPath)) {
    return { text: "", nextOffset: offset, prefixHash: "" };
  }

  try {
    const stats = fs.statSync(transcriptPath);
    const size = stats.size;
    const prefixBuffer = Buffer.alloc(Math.min(128, size));
    const handle = fs.openSync(transcriptPath, "r");

    try {
      let prefixBytes = 0;
      if (prefixBuffer.length > 0) {
        prefixBytes = fs.readSync(handle, prefixBuffer, 0, prefixBuffer.length, 0);
      }

      const prefixHash =
        prefixBytes > 0
          ? crypto.createHash("sha256").update(prefixBuffer.subarray(0, prefixBytes)).digest("hex")
          : "";
      const rotated = Boolean(knownPrefixHash && prefixHash && prefixHash !== knownPrefixHash);
      const start = rotated || size < offset ? 0 : offset;
      if (size === start) {
        return { text: "", nextOffset: size, prefixHash };
      }

      const deltaSize = size - start;
      const deltaBuffer = Buffer.alloc(deltaSize);
      const readBytes = fs.readSync(handle, deltaBuffer, 0, deltaSize, start);
      const text = deltaBuffer.subarray(0, readBytes).toString("utf8");
      return { text, nextOffset: size, prefixHash };
    } finally {
      fs.closeSync(handle);
    }
  } catch {
    return { text: "", nextOffset: offset, prefixHash: knownPrefixHash };
  }
}

function buildSignalText(payload, state) {
  const parts = [];
  if (isString(payload.last_assistant_message)) {
    parts.push(payload.last_assistant_message);
  }

  const transcriptPath = isString(payload.transcript_path) ? payload.transcript_path : null;
  const { text, nextOffset, prefixHash } = readTranscriptDelta(transcriptPath, state.offset, state.transcript_prefix_hash);
  if (text) {
    parts.push(text);
  }

  return {
    text: parts.join("\n"),
    nextOffset,
    prefixHash,
  };
}

function responseForContinue(kind) {
  return { decision: "block", reason: CONTINUE_MESSAGES[kind] };
}

function responseForCap(kind, state) {
  const used = kind === "transient_rate_limit"
    ? state.attempts_rate_limit
    : kind === "transient_overload"
      ? state.attempts_overload
      : state.attempts_usage_limit;
  const cap = MAX_ATTEMPTS_BY_KIND[kind];
  return {
    continue: false,
    stopReason: `${kind} retries exhausted (${used}/${cap}; total ${state.attempts_total}/${MAX_ATTEMPTS_TOTAL}).`,
    systemMessage: STOP_MESSAGES[kind],
  };
}

function buildEventBase(payload, sessionKey, state) {
  return {
    occurred_at: new Date().toISOString(),
    session_key: sessionKey,
    turn_id: isString(payload.turn_id) ? payload.turn_id : null,
    model: isString(payload.model) ? payload.model : null,
    cwd: isString(payload.cwd) ? payload.cwd : null,
    transcript_present: isString(payload.transcript_path),
    attempts_total_after: state.attempts_total,
    attempts_kind_after:
      state.last_matched_kind === "transient_rate_limit"
        ? state.attempts_rate_limit
        : state.last_matched_kind === "transient_overload"
          ? state.attempts_overload
          : state.attempts_usage_limit,
  };
}

function updateStateForKind(state, kind) {
  const next = { ...state, last_matched_kind: kind };
  if (kind === "transient_rate_limit") {
    next.attempts_rate_limit += 1;
    next.attempts_total += 1;
  } else if (kind === "transient_overload") {
    next.attempts_overload += 1;
    next.attempts_total += 1;
  } else if (kind === "usage_limit") {
    next.attempts_usage_limit += 1;
    next.attempts_total += 1;
  }
  return next;
}

function persistStateAndEvent(db, state, event) {
  db.exec("BEGIN IMMEDIATE;");
  try {
    if (state) {
      recordHookState(db, state);
    }
    appendStopEvent(db, event);
    db.exec("COMMIT;");
  } catch (error) {
    try {
      db.exec("ROLLBACK;");
    } catch {
      // ignore
    }
    throw error;
  }
}

function coerceState(row) {
  return {
    session_key: row.session_key,
    offset: row.offset,
    attempts_total: row.attempts_total,
    attempts_rate_limit: row.attempts_rate_limit,
    attempts_overload: row.attempts_overload,
    attempts_usage_limit: row.attempts_usage_limit,
    last_processed_turn_id: row.last_processed_turn_id ?? "",
    transcript_prefix_hash: row.transcript_prefix_hash ?? "",
    migrated_from_json: row.migrated_from_json ?? 0,
    updated_at: row.updated_at,
    last_matched_kind: "no_match",
  };
}

export function processStop(payload, options = {}) {
  if (!payload || typeof payload !== "object" || Array.isArray(payload)) {
    return null;
  }

  let runtime;
  try {
    runtime = openAnalyticsDatabase(options);
  } catch {
    return null;
  }

  const { db, stateDir } = runtime;
  const sessionKey = stableSessionKey(payload);

  try {
    if (payload.stop_hook_active === true) {
      const state = coerceState(
        queryHookState(db, sessionKey) ?? {
          session_key: sessionKey,
          offset: 0,
          attempts_total: 0,
          attempts_rate_limit: 0,
          attempts_overload: 0,
          attempts_usage_limit: 0,
          last_processed_turn_id: "",
          transcript_prefix_hash: "",
          migrated_from_json: 0,
          updated_at: new Date().toISOString(),
        }
      );
      persistStateAndEvent(db, null, {
        ...buildEventBase(payload, sessionKey, state),
        matched_kind: "no_match",
        decision: "skip_active_hook",
        exhausted: 0,
        attempts_kind_after: 0,
      });
      return null;
    }

    const row = ensureHookState(db, payload, { stateDir });
    const state = coerceState(row);
    const turnId = isString(payload.turn_id) ? payload.turn_id : "";

    if (turnId && turnId === defaultString(state.last_processed_turn_id)) {
      persistStateAndEvent(db, null, {
        ...buildEventBase(payload, sessionKey, state),
        matched_kind: "no_match",
        decision: "skip_duplicate_turn",
        exhausted: 0,
        attempts_kind_after: 0,
      });
      return null;
    }

    const signal = buildSignalText(payload, state);
    const matchedKind = classifyStopText(signal.text);
    const transcriptPresent = isString(payload.transcript_path);
    const stateWithSignal = {
      ...state,
      offset: signal.nextOffset,
      transcript_prefix_hash: signal.prefixHash,
      last_processed_turn_id: turnId || state.last_processed_turn_id,
    };

    if (matchedKind === "no_match") {
      persistStateAndEvent(db, stateWithSignal, {
        ...buildEventBase(payload, sessionKey, stateWithSignal),
        matched_kind: matchedKind,
        decision: "allow_stop",
        exhausted: 0,
        attempts_kind_after: 0,
        transcript_present: transcriptPresent,
      });
      return null;
    }

    const attemptsKind = matchedKind === "transient_rate_limit"
      ? state.attempts_rate_limit
      : matchedKind === "transient_overload"
        ? state.attempts_overload
        : state.attempts_usage_limit;
    const exhausted = attemptsKind >= MAX_ATTEMPTS_BY_KIND[matchedKind] || state.attempts_total >= MAX_ATTEMPTS_TOTAL;

    if (exhausted) {
      persistStateAndEvent(db, stateWithSignal, {
        ...buildEventBase(payload, sessionKey, stateWithSignal),
        matched_kind: matchedKind,
        decision: "stop_capped",
        exhausted: 1,
        attempts_kind_after: attemptsKind,
        transcript_present: transcriptPresent,
      });
      return responseForCap(matchedKind, state);
    }

    const updated = updateStateForKind(stateWithSignal, matchedKind);
    persistStateAndEvent(db, updated, {
      ...buildEventBase(payload, sessionKey, updated),
      matched_kind: matchedKind,
      decision: "continue",
      exhausted: 0,
      attempts_kind_after:
        matchedKind === "transient_rate_limit"
          ? updated.attempts_rate_limit
          : matchedKind === "transient_overload"
            ? updated.attempts_overload
            : updated.attempts_usage_limit,
      transcript_present: transcriptPresent,
    });
    return responseForContinue(matchedKind);
  } catch {
    return null;
  } finally {
    try {
      db.close();
    } catch {
      // ignore
    }
  }
}

export function main() {
  const payload = parsePayload(readStdin());
  if (!payload) {
    return 0;
  }

  const response = processStop(payload);
  if (response) {
    process.stdout.write(JSON.stringify(response));
  }
  return 0;
}

if (process.argv[1] && path.resolve(process.argv[1]) === fileURLToPath(import.meta.url)) {
  const exitCode = main();
  process.exitCode = exitCode;
}
